src/integer/mat_z/to_string.rs

```diff
@@ -144,3 +189,46 @@ mod test_to_string {
         assert_eq!(cmp, borrowed_string);
     }
 }
+
+#[cfg(test)]
+mod test_to_utf8 {
+    use crate::integer::MatZ;
+    use std::str::FromStr;
+
+    /// Ensures that [`MatZ::to_utf8`] is inverse to [`MatZ::from_utf8`].
+    #[test]
+    fn inverse_to_from_utf8() {
+        let message = "some_random_string_1-9A-Z!?-_;:#";
+
+        let matrix = MatZ::from_utf8(message, 4, 4);
+
+        let string = matrix.to_utf8().unwrap();
+
+        assert_eq!(message, string);
+    }
+
+    /// Ensures that [`MatZ::to_utf8`] is inverse to [`MatZ::from_utf8`]
+    /// and padding is applied if necessary.
+    #[test]
+    fn inverse_incl_padding() {
+        let message = "some_random_string_1-9A-Z!?-_;";
+        let cmp_text = "some_random_string_1-9A-Z!?-_;00";
+
+        let matrix = MatZ::from_utf8(message, 4, 8);
+
+        let string = matrix.to_utf8().unwrap();
+
+        assert_eq!(cmp_text, string);
+    }
+
+    /// Ensures that [`MatZ::to_utf8`] outputs an error
+    /// if the integer contains an invalid UTF8-Encoding.
+    #[test]
+    fn invalid_encoding() {
+        // 128 is an invalid UTF8-character (at least at the end and on its own)
+        let matrix = MatZ::from_str("[[1,2],[3,128]]").unwrap();
+        let string = matrix.to_utf8();
+
+        assert!(string.is_err());
+    }
+}
```

Comment:
From what I can see some unintended behaviour can occur.

I know that we only want to use it by:
1. Create a string
2. Create a matrix

But the other direction should not be forgotten, because the other behaviour might also be worth considering:
1. Create a matrix
2. Create a string
-> here we should run into an issue: The roundtrip only works if all entries take the same number of bits, meaning that if you have an unbalanced matrix and call to_utf8, this might be unexpected from the perspective of a user if not documented well. Also I am not sure about the behaviour in general.

Also: as you are implementing it for several types and have the same function signature it might make sense to introduce a trait for this functionality

Reply:
This shouldn't be an issue. Only the `from_utf8` function works with entries that need to have the same number of bytes for each entry and it applies padding before. The `to_utf8` function doesn't require a constant number of bytes per entry. it just takes the available bytes and appends them to the existing vector. If that is a valid UTF8-Sequence, great! Otherwise, that's a problem of the developer. In this case, I don't see a reason to introduce any limitations.

Reply:
I agree that the functionality itself works, but I just wanted to say that inconsistencies might occur because of the implementation...

I want a test that instantiates a matrix with a large entry (like 128 bits or something) and one entry with a small number of bits. Then, calling to_utf8 and afterwards calling from_utf8 will most likely give you a different matrix than the original one